### ISSUE TITLE
How to create credentials with custom connection options?

### ISSUE THREAD

**Issue by banool** [grpc/grpc-node#2565]
I have this code at the moment that I use to increase the frame and window sizes:
```
class CustomChannelCred extends ChannelCredentials {
  constructor(callCredentials?: CallCredentials) {
    super();
  }

  compose(callCredentials: CallCredentials): never {
    throw new Error("Cannot compose insecure credentials");
  }

  _getConnectionOptions(): any {
    return {
      settings: {
        // This will increase the http2 frame size. Default is 16384, which is too small.
        maxFrameSize: 4194304,
        // The initial window size set here is overridden later. We need to patch the grpc-js library to allow this.
        initialWindowSize: 4194304,
        maxHeaderListSize: 8192,
        enablePush: false,
        maxConcurrentStreams: 0,
      },
    };
  }

  _isSecure(): boolean {
    return false;
  }

  _equals(other: ChannelCredentials): boolean {
    return other instanceof CustomChannelCred;
  }
}
```

Previously I was connecting to an upstream without SSL but that has changed. I can successfully connect to the upstream if I scrap all of that and just use `credentials.createSsl()`, but then I'm not setting my custom connection options.

I've tried a variety of ways to create a custom channel credentials class, composing with CallCredentials, etc. but none of that seems to work. I'm reaching for some kind of `compose` function that works with another instance of ChannelCredentials, but it doesn't exist.

In short, how can I modify the connection options on an instance of ChannelCredentials?

**Comment by murgatroid99** [comment]
Setting custom connection options is not an intended usage of the credentials API. It is just an artifact of how the TLS credentials operate.

We can discuss implementing channel arguments to set those connection options. In particular, regarding the settings you have there:

 - `maxFrameSize` has [a channel argument](https://grpc.github.io/grpc/cpp/group__grpc__arg__keys.html#ga2ef6e079e5d61b34ce236f50fd38fc83) defined. Implementing it should be simple.
 - `initialWindowSize` is discussed in #2429. The conclusion there is that the right path forward is [BDP-based dynamic flow control window adjustment](https://grpc.io/blog/grpc-go-perf-improvements/#bdp-estimation-and-dynamic-flow-control-window). I have not yet had an opportunity to actually implement that.
 - I think `maxHeaderListSize` is controlled by [this channel argument](https://grpc.github.io/grpc/cpp/group__grpc__arg__keys.html#ga914505cb73b2d79aa8f54d78d3746a94). Implementing that should also be simple.
 - `enablePush` is irrelevant because gRPC servers should never push anyway. I think we can just hardcode this to `false`.
 - `maxConcurrentStreams` is also irrelevant. gRPC servers do not create streams. I think this can also be hardcoded to `0`.

**Comment by banool** [comment]
Hahah yeah I could tell from the API that customizing the connection options was not really supported.

Being able to set some of those arguments would be nice indeed, consider this a +1 to the feature request! As it is now, I'm seeing that upgrading from 1.8 to 1.9 actually improves the perf issues we were seeing significantly so this doesn't seem to be a blocker for us anymore.

Thanks a lot for the prompt response!
